src/component.jl

```diff
@@ -484,3 +484,209 @@ function _shared_component(typedef, mod)
+    end
+    @inbounds c.data[c.group[c.indices[e.id]]] = v
+    return v
+end
+
+
+Base.length(c::GroupedComponent) = length(c.group)
+
+function Base.empty!(c::GroupedComponent)
+    empty!(c.indices)
+    empty!(c.group)
+    empty!(c.group_size)
+    empty!(c.data)
+    return c
+end
+
+
+function Base.pop!(c::GroupedComponent, e::Entity)
+    @boundscheck if !in(e, c)
+        throw(BoundsError(c, e))
+    end
+
+    @inbounds begin
+        id = c.indices[e.id]
+        g = c.group[id]
+
+        c.group[id] = c.group[end]
+        c.group_size[g] -= 1
+        pop!(c.group)
+        pop!(c.indices, e.id)
+
+        val = c.data[g]
+
+        if c.group_size[g] == 0
+            deleteat!(c.data, g)
+            deleteat!(c.group_size, g)
+            for i in eachindex(c.group)
+                if c.group[i] > g
+                    c.group[i] -= 1
+                end
+            end
+        end
+    end
+
+    return val
+end
+
+@inline Base.iterate(c::GroupedComponent, args...) = iterate(c.data, args...)
+Base.sortperm(c::GroupedComponent) = sortperm(c.group)
+
+
+macro grouped_component(typedef)
+    return esc(Overseer._grouped_component(typedef, __module__))
+end
+
+function _grouped_component(typedef, mod)
+    t = process_typedef(typedef, mod)
+    t1, tn = t 
+    return quote
+        $t1
+       	Overseer.component_type(::Type{$tn}) = Overseer.GroupedComponent
+    end
+end
+
+function make_unique!(c::GroupedComponent)
+    @inbounds begin
+        # Go through all groups and check if a later group has the same value.
+        # If it does mark the later group empty.
+        for i in eachindex(c.group)
+            g0 = c.group[i]
+            if c.group_size[g0] > 0
+                v0 = c.data[g0]
+                for j in i+1:length(c.group)
+                    g = c.group[j]
+                    if c.group_size[g] > 0 && c.data[g] == v0
+                        c.group_size[g0] += c.group_size[g]
+                        c.group_size[g] = 0
+                        c.group[j] = g0
+                    end
+                end
+            end
+        end
+
+        # Go through all groups again and replace empty groups (i.e. its data 
+        # and group_size) by the current last group. Adjust group indices 
+        # accordingly.
+        i = 1
+        while i <= length(c.group_size)
+            if c.group_size[i] == 0
+                if i == length(c.group_size)
+                    pop!(c.group_size)
+                    pop!(c.data)
+                    break
+                else
+                    N = length(c.group_size)
+                    c.group_size[i] = pop!(c.group_size)
+                    c.data[i] = pop!(c.data)
+                    for j in eachindex(c.group)
+                        c.group[j] = c.group[j] == N ? i : c.group[j]
+                    end
+                end
+            else
+                i += 1
+            end
+        end
+    end
+
+    return
+end
```

Comment:
I think this still needs some work. Currently it cleans up duplicates in `c.data` and `c.group_size` by adjusting where `c.group` points. Which means that `c.groups` will have duplicates after this. We'd have to reroute `c.indices` to be able to cleanup `c.group` but I think it's costly to do that in that function.

Reply:
Nevermind, `c.group` needs to have one value per entity. It's as cleaned up as it can be.